Add FXIOS-12857 [Swift 6 Migration] Static properties used for testing display a warning indicating that the PreviewModel is not a sendable type.

razvanlitianu opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-12857)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/28027)

## :bulb: Description
<!--- Describe your changes so the reviewer can understand the context and decisions made for your work -->
I have three options to fix this issue. The first option is to make the PreviewModel Sendable. However, this implies that I need to make every other type Sendable, including the generic ones. Additionally, Nimbus currently doesn’t generate Sendable conformance. The second option is to make it @unchecked Sendable, which can be safe since the PreviewModel will only be used in a debug mode. Alternatively, I could annotate it with @MainActor, which I did in the PR. What are your thoughts on these options?
## :movie_camera: Demos
<!-- Please upload screenshots or video demos of your work, if applicable -->
<!-- You can either use a table (best for before/after screenshots) or the <details> disclosure -->

| Before | After |
| - | - |
| <!--insert "before" image here--> | <!--insert "after" image here--> |
| <!--insert "before" image here--> | <!--insert "after" image here--> |
| <!--insert "before" image here--> | <!--insert "after" image here--> |

<details>
<summary>Demo</summary>
<!-- Shorthand image template: <img height=400 src="<URL>" /> -->
</details>

## :pencil: Checklist
- [ ] I filled in the ticket numbers and a description of my work
- [ ] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] I ensured unit tests pass and wrote tests for new code
- [ ] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)
- [ ] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review
- [ ] If needed, I updated documentation and added comments to complex code
- [ ] If needed, I added a backport comment (example `@Mergifyio backport release/v120`)

ih-codes commented:
> Could you put a screenshot of the warning you are fixing? 🙏

FYI @lmarceau I'm guessing this is one of the warnings (I added the .enableExperimentalFeature("StrictConcurrency") flag to the OnboardingKit package’s swiftSettings array.)

<img width="1206" height="280" alt="Screenshot 2025-07-17 at 10 38 31 AM" src="https://github.com/user-attachments/assets/6dac10cd-b437-43b3-8b1a-692f80e7acf7" />

P.S. Would be helpful to provide the screenshot next time too @razvanlitianu so we have the context 🙂

ih-codes reviewed (commented):
Looking at the code more closely, I'm actually pretty sure a bunch of types in the OnboardingKit should actually be marked `Sendable` (_not_ `@unchecked`). Most of your struct and enum types are immutable and therefore `Sendable` by default, but since we had to mark these types `public`, the compiler requires us to manually add `Sendable` conformance to all our types.

I tested locally and managed to remove all the errors from OnboardingService by making types explicitly `Sendable`. You will have to make sure your structs only contain immutable properties (`let` not `var`) and that your protocol properties are also for immutable conformance (`{ get }` not `{get set}`).

Here are some examples:
- <img height="300" alt="Screenshot 2025-07-17 at 12 22 40 PM" src="https://github.com/user-attachments/assets/c24f803b-0981-4672-8e34-31c6ea35d19e" />
- <img height="200" alt="Screenshot 2025-07-17 at 12 21 19 PM" src="https://github.com/user-attachments/assets/2b69a1b1-6c2a-44b8-b740-1a0b3bf2d0c0" />
- <img height="400" alt="Screenshot 2025-07-17 at 12 19 12 PM" src="https://github.com/user-attachments/assets/cf5d456c-6024-4cdb-848c-8ffb62588d0e" />

---
Like you mentioned in the PR description, we do get some problems with the generated code from Nimbus in the Client. Thanks for raising this issue. I have made a separate ticket for that: FXIOS-12866

In the meantime, for underlying Nimbus types that are actually `Sendable`, we can suppress these Nimbus warnings by doing the following:
```
// TODO: FXIOS-12866 Nimbus generated code should add `Sendable` to enums
extension OnboardingInstructionsPopupActions: @unchecked Sendable {}
```
(please add a `TODO: FXIOS-12866` in all these cases)

razvanlitianu commented:
@ih-codes This is incredible! Thank you for taking the time to test and explain this. I was planning to do it, but I was under the impression that changing too much code would scare the team. However, the models in this case make perfect sense to be Sendable in this manner. I’ve updated the code, and please let me know if you’d like anything changed. I’ve also added a small fix for some warnings from a PR I did yesterday. I hope you don’t mind.

ih-codes commented:
> @ih-codes This is incredible! Thank you for taking the time to test and explain this. I was planning to do it, but I was under the impression that changing too much code would scare the team. However, the models in this case make perfect sense to be Sendable in this manner. I’ve updated the code, and please let me know if you’d like anything changed. I’ve also added a small fix for some warnings from a PR I did yesterday. I hope you don’t mind.

No problem @razvanlitianu, I'm glad I could help your understanding! 💪 I personally find the `Sendable` PRs like this quite easy to review, even if they cross a number of files. The compiler basically self-validates them for me. If the type can conform to `Sendable` without needing extra work, I consider this a "small" change even if you have to add several `Sendable`. 👌  So it's not a concern here in my opinion, but thank you for also taking that into consideration with your PRs. 😁 And thank you also for fixing the try/catch warnings, I don't mind at all!
